File: src/template_project/data_management/stats4schools_smoking_template.py
"""Function(s) for cleaning the data set(s)."""

import pandas as pd


def clean_smoking_stats4schools(data):
    """Clean the smoking data set from stats4schools.

    Original source of the data can be found here: https://www.stem.org.uk/rxvt6.

    Args:
        data (pandas.DataFrame): The (uncleaned) stats4schools smoking data set.

    Returns:
        pandas.DataFrame: The cleaned data set.

    """
    clean = pd.DataFrame()

    clean["gender"] = _clean_gender(data["gender"])
    clean["marital_status"] = _clean_marital_status(data["marital_status"])
    clean["smoke"] = _clean_smoke(data["smoke"])
    clean["smoke_numerical"] = clean["smoke"].cat.codes
    clean["highest_qualification"] = _clean_highest_qualification(
        data["highest_qualification"],
    )
    clean["age"] = data["age"].astype(int)
    return clean


def _clean_gender(sr):
    return pd.Categorical(sr, categories=["Female", "Male"], ordered=False)


def _clean_marital_status(sr):
    return pd.Categorical(
        sr,
        categories=["Single", "Married", "Separated", "Divorced", "Widowed"],
        ordered=False,
    )


def _clean_smoke(sr):
    return pd.Categorical(sr, categories=["No", "Yes"], ordered=True)


def _clean_highest_qualification(sr):
    replace_mapping = {
        "GCSE/CSE": "GCSE/CSE or GCSE/O Level",
        "GCSE/O Level": "GCSE/CSE or GCSE/O Level",
        "Other/Sub Degree": "Other/Sub or Higher/Sub Degree",
        "Higher/Sub Degree": "Other/Sub or Higher/Sub Degree",
    }
    ordered_qualifications = [
        "No Qualification",
        "GCSE/CSE or GCSE/O Level",
        "ONC/BTEC",
        "Other/Sub or Higher/Sub Degree",
        "A Levels",
        "Degree",
    ]
    sr = sr.replace(replace_mapping)
    return pd.Categorical(sr, categories=ordered_qualifications, ordered=True)

File: src/template_project/data_management/stats4schools_smoking_template.r
# ======================================================================================
# Functions
# ======================================================================================
clean_stats4schools_smoking <- function(data) {
  clean <- data.frame(matrix(nrow = nrow(data), ncol = 0))

  clean["gender"] <- clean_gender(data$gender)
  clean["marital_status"] <- clean_marital_status(data$marital_status)
  clean["current_smoker"] <- clean_current_smoker(data$smoke)
  clean["current_smoker_numerical"] <- convert_current_smoker_to_numerical(clean$current_smoker)
  clean["highest_qualification"] <- clean_highest_qualification(data$highest_qualification)
  clean["age"] <- as.integer(data$age)
  return(clean)
}

convert_current_smoker_to_numerical <- function(current_smoker) {
  return(as.integer(as.factor(current_smoker)) - 1)
}

clean_gender <- function(sr) {
  return(factor(sr, levels = c("Male", "Female"), ordered = FALSE))
}

clean_marital_status <- function(sr) {
  return(
    factor(
      sr,
      levels = c("Single", "Married", "Divorced", "Widowed", "Separated"),
      ordered = FALSE
    )
  )
}

clean_current_smoker <- function(sr) {
  return(factor(sr, levels = c("No", "Yes"), ordered = TRUE))
}

clean_highest_qualification <- function(sr) {
  replace_mapping <- c(
    "GCSE/CSE or GCSE/O Level" = "GCSE/CSE",
    "GCSE/CSE or GCSE/O Level" = "GCSE/O Level",
    "Other/Sub or Higher/Sub Degree" = "Other/Sub Degree",
    "Other/Sub or Higher/Sub Degree" = "Higher/Sub Degree"
  )
  ordered_qualifications <- c(
    "No Qualification",
    "GCSE/CSE or GCSE/O Level",
    "ONC/BTEC",
    "Other/Sub or Higher/Sub Degree",
    "A Levels",
    "Degree"
  )
  sr <- forcats::fct_recode(sr, !!!replace_mapping)
  return(factor(sr, levels = ordered_qualifications, ordered = TRUE))
}

# ======================================================================================
# Get config from pytask
# ======================================================================================

args <- commandArgs(trailingOnly = TRUE)
path_to_yaml <- args[length(args)]
config <- yaml::yaml.load_file(path_to_yaml)

# ======================================================================================
# Main
# ======================================================================================

data <- read.csv(config[["data"]])
data <- clean_stats4schools_smoking(data)
saveRDS(data, file = config[["produces"]])
